astropy/table/index.py

```diff
@@ -0,0 +1,781 @@
+
+        Parameters
+        ----------
+        row : int
+            Row in the sliced coordinate system
+
+        Returns
+        -------
+        orig_row : int
+            Row in the original coordinate system
+        '''
+        return row if self.original else self.start + row * self.step
+
+    def find(self, key):
+        return self.sliced_coords(self.index.find(key))
+
+    def where(self, col_map):
+        return self.sliced_coords(self.index.where(col_map))
+
+    def range(self, lower, upper):
+        return self.sliced_coords(self.index.range(lower, upper))
+
+    def same_prefix(self, key):
+        return self.sliced_coords(self.index.same_prefix(key))
+
+    def sorted_data(self):
+        return self.sliced_coords(self.index.sorted_data())
+
+    def replace(self, row, col, val):
+        if not self._frozen:
+            self.index.replace(self.orig_coords(row), col, val)
+
+    def copy(self):
+        if not self.original:
+            # replace self.index with a new object reference
+            self.index = deepcopy(self.index)
+        return self.index
+
+    def insert_row(self, pos, vals, columns):
+        if not self._frozen:
+            self.copy().insert_row(self.orig_coords(pos), vals,
+                                   columns)
+
+    def get_row_specifier(self, row_specifier):
+        return [self.orig_coords(x) for x in
+                self.index.get_row_specifier(row_specifier)]
+
+    def remove_rows(self, row_specifier):
+        if not self._frozen:
+            self.copy().remove_rows(row_specifier)
+
+    def replace_rows(self, col_slice):
+        if not self._frozen:
+            self.index.replace_rows([self.orig_coords(x) for x in col_slice])
+
+    def __repr__(self):
+        if self.original:
+            return repr(self.index)
+        return 'Index slice {0} of {1}'.format(
+            (self.start, self.stop, self.step), self.index)
+
+    def __str__(self):
+        return repr(self)
+
+    def refresh(self, columns):
+        self.index.refresh(columns)
+
+    def reload(self):
+        self.index.reload()
+
+    def col_position(self, col_name):
+        return self.index.col_position(col_name)
+
+    @property
+    def columns(self):
+        return self.index.columns
+
+    @property
+    def data(self):
+        return self.index.data
+
+
+def get_index(table, table_copy):
+    '''
+    Inputs a table and some subset of its columns, and
+    returns an index corresponding to this subset or None
+    if no such index exists.
+
+    Parameters
+    ----------
+    table : `Table`
+        Input table
+    table_copy : `Table`
+        Subset of the columns in the table argument
+    '''
+    cols = set(table_copy.columns)
+    indices = set()
+    for column in cols:
+        for index in table[column].indices:
+            if set([x.info.name for x in index.columns]) == cols:
+                return index
+    return None
+
+class index_mode(object):
```

Comment:
I think it would be nice actually if this were a `Table` method, so one could use it like:

```python
with my_table.index_mode('freeze'):
    # Do stuff...
```
etc.  It makes sense to me as a method on the table since it's temporarily doing something to the table's internal state.  Making it a method wouldn't require much change.  Just leave this class as is (though maybe give it a private name like `_IndexModeContext`) and add to `Table`:

```python
def index_mode(self, mode):
   return _IndexModeContext(self, mode)
```

Thoughts?